### ISSUE TITLE
BigQuery: hyphen symbol (-) in table project name in BigQuery

### ISSUE THREAD

**Issue by SuperBo** [sqlfluff/sqlfluff#895]
In Big Query, we can query a table named "project-a.database.a" in which, project-a is project name. However, sqlfluff get parse error with these tables.

```sql
SELECT * FROM project-a.database.a
```

## Expected Behaviour

No parse error

## Observed Behaviour

[sqlfluff.core.errors.SQLParseError("Found unparsable section: '-a.database.a'")]

## Steps to Reproduce
```python
query = 'SELECT * FROM project-a.database.a'
r = Linter(dialect='bigquery').parse_string(query)
print(r.violations)
```

## Version
Include the output of `sqlfluff --version` along with your Python version
'0.4.1'

**Comment by alanmcruickshank** [comment]
ooh - I didn't realize this was possible. Are you aware of any other dialects which allow hyphens in table or schema names?

**Comment by alanmcruickshank** [comment]
for clarity, how would the following be parsed?

``` SELECT project-a.database.a.my_col FROM project-a.database.a```

Is that selecting `my_col` from `project-a.database.a` or is it performing a subtraction operation between `project` and `a.database.a.my_col`?

**Comment by SuperBo** [comment]
@alanmcruickshank : I only use BigQuery so I can't tell about other use cases, sorry :D.

When I tried your latter query, BigQuery tell me this error: `Unrecognized name: project at [1:8]`. So I think, hyphen only valid in table name expresions.

**Comment by SuperBo** [comment]
When investigating more, I found that BigQuery only support `table.col` and `col` in select columns. `database.table.col` will raise errors.

**Comment by alanmcruickshank** [comment]
> When I tried your latter query, BigQuery tell me this error: Unrecognized name: project at [1:8]. So I think, hyphen is only valid in table name expresions.

Thanks for checking this. Supporting that syntax could be quite tricky 🤔 .

As a workaround in the meantime, if you quote the reference (i.e. `"project-a"`) then it should _parse_. You might need to add a `noqa` statement to the line so that you don't get inconsistent quoting errors.

**Comment by alanmcruickshank** [comment]
> When investigating more, I found that BigQuery only support table.col and col in select columns. database.table.col will raise errors.

I wonder whether this should be a linting rule anyway? _overqualified_ references seem like they would hinder understanding even in the dialects that do support them.

**Comment by SuperBo** [comment]
> for clarity, how would the following be parsed?
> 
> ` SELECT project-a.database.a.my_col FROM project-a.database.a`
> 
> Is that selecting `my_col` from `project-a.database.a` or is it performing a subtraction operation between `project` and `a.database.a.my_col`?

Yeah, I can confirm with you that BigQuery will parse this as a subtraction between `project` and `a.database.a.my_col`. Then it will fail because name error `Unrecognized name: project`.

In this case, can we only fix the parsing rule of table reference in FromClause?

**Comment by barrywhart** [comment]
I can confirm that our BigQuery SQL always uses backquotes around table names, and that parses fine in BigQuery. It's only because of this issue that I learned the quotes aren't always required. I can probably pick up this ticket after my current one. I think this we should try and address this for 0.5.0. It's easy to work around, but I have a passion for squashing parse errors, since they tend to block effective usage of SQLFluff.

**Comment by barrachri** [comment]
👋 

similar issue, but extended to functions.


```
SELECT
    log_tbl.my_id,
    `another-gcp-project.functions.timestamp_parsing`(log_tbl.first_move) AS first_move,
    [`another-gcp-project`].functions.timestamp_parsing(log_tbl.orderdate) AS orderdate,
    another-gcp-project.functions.timestamp_parsing(log_tbl.delivery_data) AS delivery_data
FROM
    `gcp-project.data.year_2021` AS log_tbl
;
```

```
L:   3 | P:   5 | L026 | Reference
                       | '`another-gcp-project.functions.timestamp_parsing`'
                       | refers to table/view 'functions' not found in the FROM
                       | clause or found in parent subquery.
L:   3 | P:  54 |  PRS | Found unparsable section: '(log_tbl.first_move) AS
                       | first_move'
L:   4 | P:   6 | L028 | Unqualified reference '`another-gcp-project`' found in
                       | single table select which is inconsistent with previous
                       | references.
L:   4 | P:  28 |  PRS | Found unparsable section:
                       | '.functions.timestamp_parsing(log_tbl.ord...'
L:   5 | P:  12 | L006 | Operators should be preceded by a space.
L:   5 | P:  13 | L006 | Operators should be followed by a space.
L:   5 | P:  16 | L006 | Operators should be preceded by a space.
L:   5 | P:  17 | L006 | Operators should be followed by a space.
L:   5 | P:  25 | L026 | Reference 'project.functions.timestamp_parsing' refers
                       | to table/view 'functions' not found in the FROM clause
                       | or found in parent subquery.
L:   5 | P:  52 |  PRS | Found unparsable section: '(log_tbl.delivery_data) AS
                       | delivery_data...'
L:   7 | P:  37 | L031 | Avoid using aliases in join condition
```

**Comment by barrachri** [comment]
```
L:   3 | P:   5 | L026 | Reference
                       | '`another-gcp-project.functions.timestamp_parsing`'
                       | refers to table/view 'functions' not found in the FROM
                       | clause or found in parent subquery.
L:   3 | P:  54 |  PRS | Found unparsable section: '(log_tbl.first_move) AS
                       | first_move'
L:   4 | P:   6 | L028 | Unqualified reference '`another-gcp-project`' found in
                       | single table select which is inconsistent with previous
                       | references.
L:   4 | P:  28 |  PRS | Found unparsable section:
                       | '.functions.timestamp_parsing(log_tbl.ord...'
L:   5 | P:  12 | L006 | Operators should be preceded by a space.
L:   5 | P:  13 | L006 | Operators should be followed by a space.
L:   5 | P:  16 | L006 | Operators should be preceded by a space.
L:   5 | P:  17 | L006 | Operators should be followed by a space.
L:   5 | P:  25 | L026 | Reference 'project.functions.timestamp_parsing' refers
                       | to table/view 'functions' not found in the FROM clause
                       | or found in parent subquery.
L:   5 | P:  52 |  PRS | Found unparsable section: '(log_tbl.delivery_data) AS
                       | delivery_data...'
L:   7 | P:  37 | L031 | Avoid using aliases in join condition
(.venv) [~/P/w/r2d2|] ⚡️ 1.390s
‹cbarra_add_sql_linters*› »»»» sqlfluff parse transit_time_model/queries/test.sql                                                                                                                                  65|14:40:38
[0](1, 1, 1)        |file:
[0](1, 1, 1)        |    statement:
[0](1, 1, 1)        |        select_statement:
[0](1, 1, 1)        |            select_clause:
[0](1, 1, 1)        |                keyword:                                      'SELECT'
[6](1, 1, 7)        |                [META] indent:
[6](1, 1, 7)        |                newline:                                      '\n'
[7](1, 2, 1)        |                whitespace:                                   '    '
[11](1, 2, 5)       |                select_clause_element:
[11](1, 2, 5)       |                    column_reference:
[11](1, 2, 5)       |                        identifier:                           'log_tbl'
[18](1, 2, 12)      |                        dot:                                  '.'
[19](1, 2, 13)      |                        identifier:                           'my_id'
[24](1, 2, 18)      |                comma:                                        ','
[25](1, 2, 19)      |                newline:                                      '\n'
[26](1, 3, 1)       |                whitespace:                                   '    '
[30](1, 3, 5)       |                select_clause_element:
[30](1, 3, 5)       |                    column_reference:
[30](1, 3, 5)       |                        identifier:                           '`another-gcp-project.functions.timestamp_parsing`'
[79](1, 3, 54)      |                    unparsable:                               !! Expected: 'Nothing...'
[79](1, 3, 54)      |                        start_bracket:                        '('
[80](1, 3, 55)      |                        raw:                                  'log_tbl'
[87](1, 3, 62)      |                        raw:                                  '.'
[88](1, 3, 63)      |                        raw:                                  'first_move'
[98](1, 3, 73)      |                        end_bracket:                          ')'
[99](1, 3, 74)      |                        whitespace:                           ' '
[100](1, 3, 75)     |                        raw:                                  'AS'
[102](1, 3, 77)     |                        whitespace:                           ' '
[103](1, 3, 78)     |                        raw:                                  'first_move'
[113](1, 3, 88)     |                comma:                                        ','
[114](1, 3, 89)     |                newline:                                      '\n'
[115](1, 4, 1)      |                whitespace:                                   '    '
[119](1, 4, 5)      |                select_clause_element:
[119](1, 4, 5)      |                    expression:
[119](1, 4, 5)      |                        array_literal_type:
[119](1, 4, 5)      |                            start_square_bracket:             '['
[120](1, 4, 6)      |                            [META] indent:
[120](1, 4, 6)      |                            expression:
[120](1, 4, 6)      |                                column_reference:
[120](1, 4, 6)      |                                    identifier:               '`another-gcp-project`'
[141](1, 4, 27)     |                            [META] dedent:
[141](1, 4, 27)     |                            end_square_bracket:               ']'
[142](1, 4, 28)     |                    unparsable:                               !! Expected: 'Nothing...'
[142](1, 4, 28)     |                        raw:                                  '.'
[143](1, 4, 29)     |                        raw:                                  'functions'
[152](1, 4, 38)     |                        raw:                                  '.'
[153](1, 4, 39)     |                        raw:                                  'timestamp_parsing'
[170](1, 4, 56)     |                        start_bracket:                        '('
[171](1, 4, 57)     |                        raw:                                  'log_tbl'
[178](1, 4, 64)     |                        raw:                                  '.'
[179](1, 4, 65)     |                        raw:                                  'orderdate'
[188](1, 4, 74)     |                        end_bracket:                          ')'
[189](1, 4, 75)     |                        whitespace:                           ' '
[190](1, 4, 76)     |                        raw:                                  'AS'
[192](1, 4, 78)     |                        whitespace:                           ' '
[193](1, 4, 79)     |                        raw:                                  'orderdate'
[202](1, 4, 88)     |                comma:                                        ','
[203](1, 4, 89)     |                newline:                                      '\n'
[204](1, 5, 1)      |                whitespace:                                   '    '
[208](1, 5, 5)      |                select_clause_element:
[208](1, 5, 5)      |                    expression:
[208](1, 5, 5)      |                        column_reference:
[208](1, 5, 5)      |                            identifier:                       'another'
[215](1, 5, 12)     |                        binary_operator:                      '-'
[216](1, 5, 13)     |                        column_reference:
[216](1, 5, 13)     |                            identifier:                       'gcp'
[219](1, 5, 16)     |                        binary_operator:                      '-'
[220](1, 5, 17)     |                        column_reference:
[220](1, 5, 17)     |                            identifier:                       'project'
[227](1, 5, 24)     |                            dot:                              '.'
[228](1, 5, 25)     |                            identifier:                       'functions'
[237](1, 5, 34)     |                            dot:                              '.'
[238](1, 5, 35)     |                            identifier:                       'timestamp_parsing'
[255](1, 5, 52)     |                    unparsable:                               !! Expected: 'Nothing...'
[255](1, 5, 52)     |                        start_bracket:                        '('
[256](1, 5, 53)     |                        raw:                                  'log_tbl'
[263](1, 5, 60)     |                        raw:                                  '.'
[264](1, 5, 61)     |                        raw:                                  'delivery_data'
[277](1, 5, 74)     |                        end_bracket:                          ')'
[278](1, 5, 75)     |                        whitespace:                           ' '
[279](1, 5, 76)     |                        raw:                                  'AS'
[281](1, 5, 78)     |                        whitespace:                           ' '
[282](1, 5, 79)     |                        raw:                                  'delivery_data'
[295](1, 5, 92)     |            newline:                                          '\n'
[296](1, 6, 1)      |            [META] dedent:
[296](1, 6, 1)      |            from_clause:
[296](1, 6, 1)      |                keyword:                                      'FROM'
[300](1, 6, 5)      |                newline:                                      '\n'
[301](1, 7, 1)      |                whitespace:                                   '    '
[305](1, 7, 5)      |                from_expression:
[305](1, 7, 5)      |                    [META] indent:
[305](1, 7, 5)      |                    from_expression_element:
[305](1, 7, 5)      |                        table_expression:
[305](1, 7, 5)      |                            table_reference:
[305](1, 7, 5)      |                                identifier:                   '`gcp-project.data.year_2021`'
[333](1, 7, 33)     |                        whitespace:                           ' '
[334](1, 7, 34)     |                        alias_expression:
[334](1, 7, 34)     |                            keyword:                          'AS'
[336](1, 7, 36)     |                            whitespace:                       ' '
[337](1, 7, 37)     |                            identifier:                       'log_tbl'
[344](1, 7, 44)     |                    [META] dedent:
[344](1, 7, 44)     |    newline:                                                  '\n'
[345](1, 8, 1)      |    statement_terminator:                                     ';'
[346](1, 8, 2)      |    newline:                                                  '\n'

==== parsing violations ====
L:   3 | P:  54 |  PRS | Found unparsable section: '(log_tbl.first_move) AS first_move'
L:   4 | P:  28 |  PRS | Found unparsable section:
                       | '.functions.timestamp_parsing(log_tbl.ord...'
L:   5 | P:  52 |  PRS | Found unparsable section: '(log_tbl.delivery_data) AS
                       | delivery_data...'

```

**Comment by barrachri** [comment]
The syntax should be correct: https://cloud.google.com/bigquery/docs/reference/standard-sql/functions-and-operators#calling_persistent_user-defined_functions_udfs

**Comment by barrywhart** [comment]
@barrachri: When I paste your query into the BigQuery editor, I get a syntax error:
```
Syntax error: Function call cannot be applied to this expression. Function calls require a path, e.g. a.b.c() at [4:56]
```

Can you double check that BigQuery is okay with your query, and also open a separate issue for this? Even if the issues are similar, they are likely to involve changes to different parts of the BigQuery dialect. Thanks!

**Comment by barrywhart** [comment]
@alanmcruickshank, @dmateusp: I started looking at how to fix this and quickly became blocked. IIUC, it seems there is something about the lexer that is not allowing identifiers to lex with hyphens. See the small diff below -- this is not a proposed solution; this is just initial exploration.

```
diff --git a/src/sqlfluff/dialects/dialect_ansi.py b/src/sqlfluff/dialects/dialect_ansi.py
index 4182116a..ed5ad818 100644
--- a/src/sqlfluff/dialects/dialect_ansi.py
+++ b/src/sqlfluff/dialects/dialect_ansi.py
@@ -239,7 +239,7 @@ ansi_dialect.add(
     NakedIdentifierSegment=SegmentGenerator(
         # Generate the anti template from the set of reserved keywords
         lambda dialect: ReSegment.make(
-            r"[A-Z0-9_]*[A-Z][A-Z0-9_]*",
+            r"a-b",
             name="naked_identifier",
             type="identifier",
             _anti_template=r"^(" + r"|".join(dialect.sets("reserved_keywords")) + r")$",
@@ -822,12 +822,13 @@ class FromExpressionElementSegment(BaseSegment):
     """A table expression."""

     type = "from_expression_element"
-    match_grammar = Sequence(
-        Ref("PreTableFunctionKeywordsGrammar", optional=True),
-        OptionallyBracketed(Ref("TableExpressionSegment")),
-        Ref("AliasExpressionSegment", optional=True),
-        Ref("PostTableExpressionGrammar", optional=True),
-    )
+    # match_grammar = Sequence(
+    #     Ref("PreTableFunctionKeywordsGrammar", optional=True),
+    #     OptionallyBracketed(Ref("TableExpressionSegment")),
+    #     Ref("AliasExpressionSegment", optional=True),
+    #     Ref("PostTableExpressionGrammar", optional=True),
+    # )
+    match_grammar = Ref("NakedIdentifierSegment")

     def get_eventual_alias(self) -> Optional[AliasInfo]:
         """Return the eventual table name referred to by this table expression.
```

See the change to `NakedIdentifierSegment`: I've updated it so the only possible identifier is `a-b`. 😄

With these changes in place, I would expect the SQL `SELECT * FROM a-b` to parse successfully, but it fails with:
```
L:   1 | P:  10 |  PRS | Found unparsable section: 'FROM a-b'
```

If I change the `NakedIdentifierSegment` regex to something else, such as `ab`, a similar query parses successfully:
```
SELECT * FROM ab
```

And I can confirm that this hack *is* impacting the parsing of table names, because this query, as expected, does not parse (because the name is not `ab`):
```
SELECT * FROM abc
```

Any ideas why the hyphen regex is not working?

**Comment by dmateusp** [comment]
@barrywhart I gave it a try and it seems like your regex was not working because we're treating `-` as a special character in the lexer, here's a change that makes it parse:

```diff
diff --git a/src/sqlfluff/dialects/dialect_ansi.py b/src/sqlfluff/dialects/dialect_ansi.py
index 4182116a..d22ce560 100644
--- a/src/sqlfluff/dialects/dialect_ansi.py
+++ b/src/sqlfluff/dialects/dialect_ansi.py
@@ -100,7 +100,7 @@ ansi_dialect.set_lexer_struct(
         ("comma", "singleton", ",", dict(is_code=True, type="comma")),
         ("plus", "singleton", "+", dict(is_code=True)),
         ("tilde", "singleton", "~", dict(is_code=True)),
-        ("minus", "singleton", "-", dict(is_code=True)),
+        # ("minus", "singleton", "-", dict(is_code=True)),
         ("divide", "singleton", "/", dict(is_code=True)),
         ("percent", "singleton", "%", dict(is_code=True)),
         ("star", "singleton", "*", dict(is_code=True)),
@@ -112,7 +112,7 @@ ansi_dialect.set_lexer_struct(
         ("crly_bracket_close", "singleton", "}", dict(is_code=True)),
         ("colon", "singleton", ":", dict(is_code=True)),
         ("semicolon", "singleton", ";", dict(is_code=True)),
-        ("code", "regex", r"[0-9a-zA-Z_]*", dict(is_code=True)),
+        ("code", "regex", r"[0-9a-zA-Z_\-]*", dict(is_code=True)),
     ]
 )
 
@@ -239,7 +239,7 @@ ansi_dialect.add(
     NakedIdentifierSegment=SegmentGenerator(
         # Generate the anti template from the set of reserved keywords
         lambda dialect: ReSegment.make(
-            r"[A-Z0-9_]*[A-Z][A-Z0-9_]*",
+            r"[A-Z0-9_]*-?[A-Z][A-Z0-9_]*",
             name="naked_identifier",
             type="identifier",
             _anti_template=r"^(" + r"|".join(dialect.sets("reserved_keywords")) + r")$",
```

This isn't a solution, I imagine we would need to change the NakedIdentifierSegment to allow an identifier that optionally includes a MinusSegment

**Comment by barrywhart** [comment]
Thanks, @NiallRees -- that was a big help. I wasn't aware of the `code` regex, which I guess was blocking hyphens from ever making it to the `NakedIdentifierSegment` regex.

I have some ideas that I'm exploring and will put up a PR or patch, depending on how promising it looks. As Alan pointed out, we likely need two different flavors of segment for BigQuery:
* One following `FROM` which allows naked (i.e. unquoted) names that contain a hyphen
* The traditional `NakedIdentifierSegment` for use in all the other places, since hyphens normally mean subtraction

It's odd that BigQuery even allows this, but they do, so SQLFluff should also support it. 😬

**Comment by barrywhart** [comment]
I did some experimenting with the UDF example in this issue. Based on the following, I don't think BigQuery parses hyphenated "naked" project names correctly.

Query:
```
SELECT
    another-gcp-project.functions.timestamp_parsing(1)
```

Running this query results in the error:
```
Not found: Project project
```

Apparently, each part of the hyphenated name is being parsed separately. If I change the query to quote the project name, i.e.:
```
SELECT
    `another-gcp-project`.functions.timestamp_parsing(1)
```

I get a different error (because the project doesn't exist), but the error indicates the project name is being parsed correctly now:
```
Not found: Project another-gcp-project
```

There may indeed be an issue with UDFs, but as I mentioned earlier, this is separate enough from the original issue that I suggest creating another issue. It would be cool to have a more complete example that parses and runs in BigQuery but is not processed correctly by SQLFluff. For tricky issues like this one, it's difficult to properly triage and analyze it with non-working code snippets.

**Comment by barrachri** [comment]
Thanks @barrywhart, I opened a new issue: https://github.com/sqlfluff/sqlfluff/issues/911

BTW, sqlfluff is pretty cool!

**Comment by barrywhart** [comment]
Thanks for the kind words, @barrachri! I started working on the project because it's useful, but most of the work I do now is because it's fun. 🎉
